lib/charms/prometheus_k8s/v0/prometheus.py

```diff
@@ -375,6 +425,64 @@ def jobs(self):
 
         return scrape_jobs
 
+    def alerts(self):
+        """Fetch alerts for all relations.
+
+        A Prometheus alert rules file consists of a list of "groups". Each
+        group consists of a list of alerts (`rules`) that are sequentially
+        executed. This method returns all the alert rules provided by each
+        related consumer charm. These rules may be used to generate a
+        separate alert rules file for each relation since the returned list
+        of alert groups are indexed by relation ID. Also for each relation ID
+        associated scrape metadata such as Juju model, UUID and application
+        name are provided so the a unique name may be generated for the rules
+        file. For each relation the structure of data returned is a dictionary
+        with four keys
+
+        - groups
+        - model
+        - model_uuid
+        - application
+
+        The value of the `groups` key is such that it may be used to generate
+        a Prometheus alert rules file directly using `yaml.dump` but the
+        `groups` key itself must be included as this is required by Prometheus,
+        for example as in `yaml.dump({"groups": alert["groups"]})`.
+
+        Currently the consumer charm only accepts a list of rules and these
+        rules are all placed into a single group, even though Prometheus itself
+        allows for multiple groups within a single alert rules file.
+
+        Returns:
+            a dictionary of alert rule groups and associated scrape
+            metadata indexed by relation ID.
+        """
+        alerts = {}
+        for relation in self._charm.model.relations[self._relation_name]:
+            if not relation.units:
+                continue
+
+            alert_rules = json.loads(
+                relation.data[relation.app].get("alert_rules", "{}")
+            )
+
+            scrape_metadata = json.loads(
+                relation.data[relation.app].get("scrape_metadata", "{}")
+            )
+
+            if alert_rules and scrape_metadata:
+                try:
+                    alerts[relation.id] = {
+                        "groups": alert_rules["groups"],
+                        "model": scrape_metadata["model"],
+                        "model_uuid": scrape_metadata["model_uuid"][:7],
+                        "application": scrape_metadata["application"],
+                    }
+                except KeyError:
+                    logger.error("Relation %s has invalid data", relation.id)
```

Comment:
Hmm, what happens when if this exception fires? Do you have a half formed alert in the alerts list, or does it fail to assign completely?


below is ugly but trying to convey the point

```suggestion
           valid = ("groups" in alert_rules.keys() and 
                          all(i in ["model","model_uuid","application"] for i in scrape_metadata.keys()))
            
            if valid:
                alerts[relation.id] = {
                    "groups": alert_rules["groups"],
                    "model": scrape_metadata["model"],
                    "model_uuid": scrape_metadata["model_uuid"][:7],
                    "application": scrape_metadata["application"],
                }
```